src/session_server_ssh.c

```diff
@@ -176,6 +177,41 @@ nc_server_ssh_set_interactive_auth_clb(int (*interactive_auth_clb)(const struct
     server_opts.interactive_auth_data_free = free_user_data;
 }
 
+API int
+nc_server_ssh_set_pam_conf_path(const char *conf_name, const char *conf_dir)
+{
+    int r = 0;
+
+    free(server_opts.conf_name);
+    free(server_opts.conf_dir);
+    server_opts.conf_name = NULL;
+    server_opts.conf_dir = NULL;
+
+    if (conf_name) {
+        server_opts.conf_name = strdup(conf_name);
+        if (!(server_opts.conf_name)) {
+            ERRMEM;
+            r = -1;
+        }
+    }
+
+    if (conf_dir) {
+#ifdef LIBPAM_HAVE_CONFDIR
+        server_opts.conf_dir = strdup(conf_dir);
+        if (!(server_opts.conf_dir)) {
+            ERRMEM;
+            r = -1;
+        }
+#else
+        ERR(NULL, "Failed to set PAM config directory because of old version of PAM. "
+                "Put the config file in the system directory (usually /etc/pam.d/).");
+        r = -1;
```

Comment:
In case user sets both `conf_name` and `conf_dir` and this code is executed, the return value would be -1 despite `conf_name` being set. Such behavior should be avoided, the general rule (that has almost no exceptions) is that either function fully succeeds or fails without changing anything. Partial success/fail is confusing and usually not intentional.

Reply:
`conf_dir` is now set first. This means that if it fails, `conf_name` remains unchanged.